### DISCUSSION TITLE
FATA[0003] pids cgroup controller not found

### DISCUSSION THREAD

**Discussion by JoCa96** [k3s-io/k3s#7025] [Q&A] [answered]
**Environmental Info:**
K3s Version: 
rancher/k3s:v1.26.2-rc1-k3s1-arm-data-1

Node(s) CPU architecture, OS, and Version: 
Linux my-nas 4.2.8 #2 SMP Thu Jan 12 10:34:18 CST 2023 armv7l unknown

Cluster Configuration:
1 servers, 2 agents

**Describe the bug:**
I am trying to create a k3s node on my QNAP NAS using the "Container Station" (= QNAP Docker runtime).

Unfortunately the container crashes with the error `FATA[0003] pids cgroup controller not found`.

On the host system the followng cgroups are mounted:
```sh
[~] # cat /proc/mounts | grep cgroup
cgroup_root /sys/fs/cgroup tmpfs rw,relatime 0 0
none /sys/fs/cgroup/memory cgroup rw,relatime,memory 0 0
cpu /sys/fs/cgroup/cpu cgroup rw,relatime,cpu 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,relatime,cpuset 0 0
cgroup /sys/fs/cgroup/cpuacct cgroup rw,relatime,cpuacct 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,relatime,blkio 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,relatime,devices 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,relatime,freezer 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,relatime,name=systemd 0 0
```

This is probably **not** a bug, but I don't know how I can resolve this issue as I am unable to find any information related to the error `FATA[0003] pids cgroup controller not found`.

**Expected behavior:**
The container start successfully.

**Actual behavior:**
FATA[0003] pids cgroup controller not found

**Additional context / logs:**
This is the start-up log of the container:
```
WARN[0000] cgroup2 is not mounted. cgroup2 evacuation is discarded. 
INFO[0000] Starting k3s agent v1.26.2-rc1+k3s1 (ea094d1d) 
INFO[0000] Running load balancer k3s-agent-load-balancer 127.0.0.1:6444 -> [x.x.x.x:6443] 
WARN[0002] Failed to load kernel module overlay with modprobe 
INFO[0002] Module nf_conntrack was already loaded       
INFO[0002] Module br_netfilter was already loaded       
INFO[0002] Module iptable_nat was already loaded        
WARN[0002] Failed to load kernel module iptable_filter with modprobe 
W0305 16:26:15.037751      17 sysinfo.go:203] Nodes topology is not available, providing CPU topology
INFO[0002] Logging containerd to /var/lib/rancher/k3s/agent/containerd/containerd.log 
INFO[0002] Running containerd -c /var/lib/rancher/k3s/agent/etc/containerd/config.toml -a /run/k3s/containerd/containerd.sock --state /run/k3s/containerd --root /var/lib/rancher/k3s/agent/containerd 
INFO[0003] containerd is now running                    
INFO[0003] Getting list of apiserver endpoints from server 
INFO[0003] Updating load balancer k3s-agent-load-balancer default server address -> x.x.x.x:6443 
INFO[0003] Connecting to proxy                           url="wss://x.x.x.x:6443/v1-k3s/connect"
WARN[0003] Disabling CPU quotas due to missing cpu controller or cpu.cfs_period_us 
FATA[0003] pids cgroup controller not found             
```

**Comment by brandond** [comment]
Some cgroups are not optional; the Kubelet cannot function without them. It appears that the kernel on your appliance lacks support for this cgroup. Another possibility is that it uses hybrid cgroup v2, with some controllers bound to v2 and some still on v1. This is not supported either.

**Reply by brandond** [reply]
I'm not familiar with qnap's version of Docker, but you might also confirm that you're running the k3s container privileged?

**Reply by JoCa96** [reply]
Thanks for your help!
The container is already running in privileged mode.

I have root access to the device, is it possible to just enable this cgroup or is that only possible at kernel compilation time?

**Reply by brandond** [reply, marked as answer]
You can run `cat /proc/cgroups` to see which cgroups are supported. If any of the required cgroups are missing, it is most likely that the manufacturer has not enabled those cgroups in the custom kernel that they're building for your device.

**Reply by JoCa96** [reply]
Seems like it is not supported. But thanks for your help!
